pkg/controllers/management/globaldns/globaldns_provider.go

```diff
@@ -139,6 +143,47 @@ func (n *ProviderLauncher) handleRoute53Provider(obj *v3.GlobalDNSProvider) (run
 	return nil, nil
 }
 
+func (n *ProviderLauncher) handleCloudflareProvider(obj *v3.GlobalDNSProvider) (runtime.Object, error) {
+	//create external-dns route53 provider
+	data := map[string]interface{}{
+		"apiKey":           obj.Spec.CloudflareProviderConfig.APIKey,
+		"apiEmail":         obj.Spec.CloudflareProviderConfig.APIEmail,
+		"cloudflareDomain": obj.Spec.CloudflareProviderConfig.RootDomain,
+		"deploymentName":   obj.Name,
+	}
+	cfTemplate := template.Must(template.New("cloudflare_template").Parse(CloudflareDeploymentTemplate))
+	var output bytes.Buffer
+	if err := cfTemplate.Execute(&output, data); err != nil {
+		return nil, fmt.Errorf("GlobaldnsProviderLauncher: Error parsing the external-dns/cloudflare deployment template: %v", err)
+	}
+
+	decode := scheme.Codecs.UniversalDeserializer().Decode
+	deployObj, _, err := decode(output.Bytes(), nil, nil)
+	if err != nil {
+		return nil, fmt.Errorf("GlobaldnsProviderLauncher: Error decoding  external-dns/cloudflare deployment template to Kubernetes Object: %v", err)
+	}
+
+	deployment := deployObj.(*v1beta1.Deployment)
+	//set ownerRef to the dnsProvider CR
+	controller := true
+	ownerRef := []metav1.OwnerReference{{
+		Name:       obj.Name,
+		APIVersion: "v3",
+		UID:        obj.UID,
+		Kind:       obj.Kind,
+		Controller: &controller,
+	}}
+	deployment.ObjectMeta.OwnerReferences = ownerRef
+
+	deploymentCreated, err := n.Deployments.Create(deployment)
+	if err != nil {
+		return nil, fmt.Errorf("GlobaldnsProviderLauncher: Error creating external-dns deployment for Cloudflare provider: %v ", err)
+	}
+	logrus.Infof("GlobaldnsProviderLauncher: Created Cloudflare external-dns provider deployment %v", deploymentCreated.Name)
+
+	return nil, nil
+}
+
```

Comment:
Can we take out all the template parsing/deployment code that is common to route53 and cloudflare into a common method and reuse for both providers?